File: src/wiki/render.py
"""Dynamic SPARQL block rendering and wikilink-to-HTML conversion."""

from __future__ import annotations

from pathlib import Path
import re
from typing import Any

import click
from markdown_it import MarkdownIt

from .format import run_query
from wiki.mdit_py_plugins.wikilink import wikilink_plugin
from .paths import iter_markdown_files, page_routes, select_markdown_paths

# Matches SPARQL wrapper comments, fenced query, rendered table, and end comment.
SPARQL_BLOCK_REGEX = re.compile(
    r"(?P<start><!--\s*sparql:start\s*-->)(?P<prefix>\s*)"
    r"(?P<fence>```sparql\s*(?P<query>.*?)\s*```)(?P<mid>\s*)"
    r"(?P<table>.*?)(?P<suffix>\s*)(?P<end><!--\s*sparql:end\s*-->)",
    re.DOTALL | re.IGNORECASE,
)


def has_sparql_blocks(md_file: Path) -> bool:
    """Return True if the markdown file contains at least one SPARQL block."""
    try:
        content = md_file.read_text(encoding="utf-8")
    except OSError:
        return False
    return SPARQL_BLOCK_REGEX.search(content) is not None


def select_markdown_files_for_render(
    context: Any,
    *,
    explicit_files: tuple[Path, ...] = (),
) -> list[Path]:
    """Choose markdown files to process for SPARQL rendering."""
    if explicit_files:
        candidates = select_markdown_paths(context, explicit_files)
    else:
        candidates = iter_markdown_files(context)
    return [md_file for md_file in candidates if has_sparql_blocks(md_file)]


def _replace_sparql_table(match: re.Match[str], rendered_markdown: str) -> str:
    """Rebuild a SPARQL block, preserving the on-disk query fence and surrounding whitespace."""
    suffix = match.group("suffix")
    if not suffix:
        suffix = "\n"
    return (
        f"{match.group('start')}{match.group('prefix')}{match.group('fence')}"
        f"{match.group('mid')}{rendered_markdown}{suffix}{match.group('end')}"
    )


def _is_table_divider_row(cells: list[str]) -> bool:
    return bool(cells) and all(re.fullmatch(r"-+", cell) for cell in cells)


def _normalize_markdown_table(text: str) -> str:
    """Normalize GFM table padding so compact and mdformat-padded tables compare equal."""
    lines = [line.strip() for line in text.strip().splitlines() if line.strip()]
    normalized: list[str] = []
    for index, line in enumerate(lines):
        if not line.startswith("|"):
            normalized.append(line)
            continue
        cells = [cell.strip() for cell in line.strip("|").split("|")]
        if index == 0:
            cells = [cell.casefold() for cell in cells]
        elif _is_table_divider_row(cells):
            cells = ["---"] * len(cells)
        normalized.append("| " + " | ".join(cells) + " |")
    return "\n".join(normalized)


def _sparql_table_matches(existing: str, rendered: str) -> bool:
    """Return True when the on-disk table already matches the rendered output."""
    return _normalize_markdown_table(existing) == _normalize_markdown_table(rendered)


def render_markdown_files(
    context: Any,
    graph: Any,
    dry_run: bool = False,
    explicit_files: tuple[Path, ...] = (),
) -> tuple[int, int, list[str]]:
    """Iterate over markdown files, parse and replace dynamic SPARQL sections inline.

    Returns (success_count, error_count, stale_files).
    """
    success_count = 0
    error_count = 0
    stale_files: list[str] = []
    markdown_files = select_markdown_files_for_render(context, explicit_files=explicit_files)

    known_slugs = {pr.route for pr in page_routes(context)}

    for md_file in markdown_files:
        content = md_file.read_text(encoding="utf-8")
        modified = False
        file_errors = 0

        def replacer(match: re.Match[str]) -> str:
            nonlocal modified, file_errors
            query = match.group("query").strip()
            try:
                rendered_markdown = run_query(
                    graph,
                    query,
                    output_format="markdown",
                    wiki_base=context.wiki_base,
                    known_slugs=known_slugs,
                )
                existing_table = match.group("table")
                if _sparql_table_matches(existing_table, rendered_markdown):
                    return match.group(0)
                modified = True
                return _replace_sparql_table(match, rendered_markdown)
            except Exception as e:
                click.echo(f"Error rendering query in {md_file.name}: {e}", err=True)
                file_errors += 1
                return match.group(0)

        new_content = SPARQL_BLOCK_REGEX.sub(replacer, content)
        if modified:
            if new_content == content:
                continue
            try:
                rel = str(md_file.relative_to(Path.cwd()))
            except ValueError:
                rel = str(md_file)
            stale_files.append(rel)

            if not dry_run:
                md_file.write_text(new_content, encoding="utf-8")
                success_count += 1
        error_count += file_errors

    return (success_count, error_count, stale_files)


def render_markdown(text: str) -> str:
    """Render wikilinks in *text* to HTML anchor tags, leaving all other markdown intact."""
    md = MarkdownIt("gfm-like", {"linkify": False})
    md.use(wikilink_plugin)
    return md.render(text)

File: src/wiki/render.ts
import type { RdfDataset, RdfGraph } from "./rdf.ts";
import { readTextTolerant } from "./parser.ts";
import { iterMarkdownFiles, selectMarkdownPaths } from "./paths.ts";
import type { Path } from "./fspath.ts";
import { runQuery } from "./format.ts";
import type { Config } from "./config.ts";
import type { RenderReport } from "./schemas/reports.ts";

const SPARQL_BLOCK_REGEX =
  /(?<start><!--\s*sparql:start(?:\s*-->)?)(?<prefix>\s*)(?<fence>```sparql\s*(?<query>.*?)\s*```)(?<mid>\s*)(?:(?<comment_end>-->)(?<post_comment>\s*))?(?<table>.*?)(?<suffix>\s*)(?<end><!--\s*sparql:end\s*-->)/gis;

function blockRegex(): RegExp {
  return new RegExp(SPARQL_BLOCK_REGEX.source, SPARQL_BLOCK_REGEX.flags);
}

function hasSparqlBlocks(path: Path): boolean {
  try {
    return blockRegex().test(readTextTolerant(path));
  } catch {
    return false;
  }
}

function normalizeMarkdownTable(text: string): string {
  const lines = text.trim().split(/\r?\n/).map((line) => line.trim()).filter(
    Boolean,
  );
  const normalized: string[] = [];
  for (let index = 0; index < lines.length; index += 1) {
    const line = lines[index]!;
    if (!line.startsWith("|")) {
      normalized.push(line);
      continue;
    }
    const cells = line.replace(/^\|+/, "").replace(/\|+$/, "").split("|").map((
      cell,
    ) => cell.trim());
    if (index === 0) {
      normalized.push(
        "| " + cells.map((cell) => cell.toLowerCase()).join(" | ") + " |",
      );
    } else if (cells.length > 0 && cells.every((cell) => /^-+$/.test(cell))) {
      normalized.push("| " + cells.map(() => "---").join(" | ") + " |");
    } else {
      normalized.push("| " + cells.join(" | ") + " |");
    }
  }
  return normalized.join("\n");
}

function sparqlTableMatches(existing: string, rendered: string): boolean {
  return normalizeMarkdownTable(existing) === normalizeMarkdownTable(rendered);
}

function group(match: RegExpMatchArray, name: string): string {
  return match.groups?.[name] ?? "";
}

function replaceSparqlTable(
  match: RegExpMatchArray,
  renderedMarkdown: string,
): string {
  const suffix = group(match, "suffix") || "\n";
  return `${group(match, "start")}${group(match, "prefix")}${
    group(match, "fence")
  }${group(match, "mid")}${group(match, "comment_end")}${
    group(match, "post_comment")
  }${renderedMarkdown}${suffix}${group(match, "end")}`;
}

function relativePath(path: Path): string {
  try {
    return path.relativeTo(Deno.cwd()).asPosix();
  } catch {
    return path.toString();
  }
}

export interface RenderMarkdownOptions {
  readonly queryGraph: (query: string) => Promise<RdfGraph | RdfDataset>;
  readonly baseIri: string;
  readonly knownSlugs: ReadonlySet<string>;
  readonly dryRun?: boolean;
  readonly explicitFiles?: readonly Path[];
}

export async function renderMarkdownFiles(
  config: Config,
  options: RenderMarkdownOptions,
): Promise<RenderReport> {
  const explicitFiles = options.explicitFiles ?? [];
  const candidates = explicitFiles.length > 0
    ? selectMarkdownPaths(config, explicitFiles)
    : iterMarkdownFiles(config);
  const markdownFiles = candidates.filter(hasSparqlBlocks);
  const staleFiles: string[] = [];
  const renderErrors: string[] = [];
  let updatedCount = 0;
  let errorCount = 0;

  for (const file of markdownFiles) {
    const content = readTextTolerant(file);
    const matches = [...content.matchAll(blockRegex())];
    const parts: string[] = [];
    let cursor = 0;
    let modified = false;
    for (const match of matches) {
      const index = match.index;
      if (index === undefined) continue;
      parts.push(content.slice(cursor, index));
      const query = group(match, "query").trim();
      try {
        const graph = await options.queryGraph(query);
        const renderedMarkdown = await runQuery(graph, query, {
          format: "markdown",
          baseIri: options.baseIri,
          knownSlugs: options.knownSlugs,
        });
        if (sparqlTableMatches(group(match, "table"), renderedMarkdown)) {
          parts.push(match[0]);
        } else {
          modified = true;
          parts.push(replaceSparqlTable(match, renderedMarkdown));
        }
      } catch (error) {
        renderErrors.push(
          `Error rendering query in ${file.name}: ${
            error instanceof Error ? error.message : String(error)
          }`,
        );
        errorCount += 1;
        parts.push(match[0]);
      }
      cursor = index + match[0].length;
    }
    if (matches.length === 0) continue;
    parts.push(content.slice(cursor));
    const newContent = parts.join("");
    if (!modified || newContent === content) continue;
    staleFiles.push(relativePath(file));
    if (!options.dryRun) {
      await Deno.writeTextFile(file.toString(), newContent);
      updatedCount += 1;
    }
  }

  return {
    ok: options.dryRun ? staleFiles.length === 0 : errorCount === 0,
    updated_count: updatedCount,
    error_count: errorCount,
    stale_files: staleFiles,
    render_errors: renderErrors,
  };
}
